homeassistant/components/sensor/magicseaweed.py

```diff
@@ -0,0 +1,258 @@
+
+    if CONF_UNITS in config:
+        units = config.get(CONF_UNITS)
+    elif hass.config.units.is_metric:
+        units = 'eu'
+    else:
+        units = 'us'
+
+    forecast_data = MagicSeaweedData(
+        api_key=api_key,
+        spot_id=spot_id,
+        units=units,
+        interval=config.get(CONF_UPDATE_INTERVAL))
+    forecast_data.update()
+
+    # If connection failed don't setup platform.
+    if forecast_data.currently is None or forecast_data.hourly is None:
+        return False
+
+    sensors = []
+    for variable in config[CONF_MONITORED_CONDITIONS]:
+        if 'forecast' in variable:
+            sensors.append(MagicSeaweedSensor(forecast_data, variable, name,
+                                              units))
+
+        else:
+            sensors.append(MagicSeaweedSensor(forecast_data, variable, name,
+                                              units))
+            if hours is not None:
+                for hour in hours:
+                    sensors.append(MagicSeaweedSensor(
+                        forecast_data, variable, name, units, hour))
+    add_devices(sensors, True)
+
+
+class MagicSeaweedSensor(Entity):
+    """Implementation of a MagicSeaweed sensor."""
+
+    def __init__(self, forecast_data, sensor_type, name, unit_system,
+                 hour=None):
+        """Initialize the sensor."""
+        self.client_name = name
+        self._name = SENSOR_TYPES[sensor_type][0]
+        self.data = forecast_data
+        self.type = sensor_type
+        self.hour = hour
+        self._unit_system = unit_system
+        self._state = None
+        self._icon = None
+        self._unit_of_measurement = None
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        if self.hour is None and 'forecast' in self.type:
+            return "{} {}".format(self.client_name, self._name)
+        elif self.hour is None:
+            return "Current {} {}".format(self.client_name, self._name)
+        return "{} {} {}".format(
+            self.hour, self.client_name, self._name)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_system(self):
+        """Return the unit system of this entity."""
+        return self._unit_system
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def icon(self):
+        """Return the entity weather icon, if any."""
+        return ICON
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self.type == 'swell_forecast' and self.hour is None:
+            summaries = {ATTR_ATTRIBUTION: CONF_ATTRIBUTION}
+            for hour, data in self.data.hourly.items():
+                occurs = hour
+                swell = data.get('swell')
```

Comment:
Common practice is to leave the `device_state_attributes` property pretty simple – e.g.:

```
@property
    def device_state_attributes(self):
    """Return the state attributes."""
    return self._attrs
```

...and do all of the logic to update `self._attrs` in `update()`. This way, the property itself is easier to understand.